### PR TITLE
feat: task and use_task for better asyncio task a thread execution

### PR DESCRIPTION
**Pull request by maartenbreddels**
use_task can replace use_thread, but also support coroutine functions.

task is a decorator that creates a top level task that can be executed
from any location (e.g. a click handler of a button).

### CODE REVIEW

**Comment by maartenbreddels** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/widgetti/solara/461?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #452** <a href="https://app.graphite.dev/github/pr/widgetti/solara/452?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #453** <a href="https://app.graphite.dev/github/pr/widgetti/solara/453?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #454** <a href="https://app.graphite.dev/github/pr/widgetti/solara/454?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #455** <a href="https://app.graphite.dev/github/pr/widgetti/solara/455?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #461** <a href="https://app.graphite.dev/github/pr/widgetti/solara/461?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/widgetti/solara/461?utm_source=stack-comment).

File: solara/website/pages/api/task.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+The task callable does will only execute the function once when called multiple times,
+and will cancel previous executions if the function is called again before the previous finished.
+
+
+The wrapped function return value is available as the `.value` attribute of the task object.
+
+
+
+
+"""
```
Comment (iisakkirotko) [inline solara/website/pages/api/task.py:16]:
```suggestion
"""# Task

A global way to run code in the background, with the UI available to the user. This is useful for long running tasks, like downloading data.

"""
```

Currently this duplicates the docstring

File: solara/website/pages/api/use_task.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+The task callable does will only execute the function once when called multiple times,
+and will cancel previous executions if the function is called again before the previous finished.
+
+
+The wrapped function return value is available as the `.value` attribute of the task object.
+
+
+
+
+"""
```
Comment (iisakkirotko) [inline solara/website/pages/api/use_task.py:16]:
```suggestion
"""# use_task

A hook that allows you to run code in the background, with the UI available to the user. This is useful for long running tasks, like downloading data.

Note that unlike with the `@task` decorator, the result is not globally shared, but only available to the component that called `use_task`.

"""
```

Same as above.

File: solara/website/pages/docs/content/10-howto/30-tasks.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+countries = ['Aruba', 'the Netherlands', 'USA', 'China']
+country = solara.reactive('Aruba')
+
+@solara.component
+def Page():
+    def get_country_data():
+        return requests.get(f"https://restcountries.com/v3.1/name/{country.value}").json()[0]
+
+    result = use_task(get_country_data, call=True, dependencies=[country.value])
```
Comment (iisakkirotko) [inline solara/website/pages/docs/content/10-howto/30-tasks.md:76]:
```suggestion
    result = use_task(get_country_data, dependencies=[country.value])
```

Removed in new version.

File: solara/tasks.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+R = TypeVar("R")
+
+P = typing_extensions.ParamSpec("P")
+logger = logging.getLogger("solara.task")
+
+
+class Task(Generic[P, R], abc.ABC):
+    def __init__(self):
+        self.result = solara.reactive(
+            # cast(Optional[solara.Result[R]], None)
```
Comment (iisakkirotko) [inline solara/tasks.py:33]:
```suggestion
```

File: solara/tasks.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+            solara.Result[R](
+                value=None,
+                state=solara.ResultState.INITIAL,
+            )
+        )
+        # self.thread_pool = ThreadPoolExecutor(max_workers=1)
```
Comment (iisakkirotko) [inline solara/tasks.py:39]:
```suggestion
```

File: solara/tasks.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+        self.last_value: Optional[R] = None
+
+    def reset(self):
+        self.running_thread = None
+        self.cancel()
+        # self.result.value = None
```
Comment (iisakkirotko) [inline solara/tasks.py:45]:
```suggestion
```

File: solara/tasks.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> None:
+        task_for_this_call = None
+        previous_task = self.current_task
+
+        def retry():
+            self.__call__(*args, **kwargs)
+            # nonlocal task_for_this_call
+            # task_for_this_call.cancel()
+            # self.current_task = task_for_this_call = asyncio.create_task(runner())
```
Comment (iisakkirotko) [inline solara/tasks.py:93]:
```suggestion
```

File: solara/tasks.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+
+        def cancel():
+            assert task_for_this_call is not None
+            # assert asyncio.current_task() != task_for_this_call
```
Comment (iisakkirotko) [inline solara/tasks.py:97]:
```suggestion
```

**Review by iisakkirotko — CHANGES_REQUESTED** [review]
This feature is definitely going to have some demand! I'll give a possible different example in the docs some more thought.

**Pull request merged**